### ISSUE TITLE
Support for Windows on ARM 64-bit

### ISSUE THREAD

**Issue by xanderberkhout** [JFormDesigner/FlatLaf#443]
I just tested a Java Swing application with the FlatLaf on Windows 11 ARM 64-bit and I got the following error:

```
java.lang.UnsatisfiedLinkError: C:\Users\xander\AppData\Local\Temp\flatlaf.temp\flatlaf-windows-x86-20900948333330.dll: Can't load IA 32-bit .dll on a ARM 64-bit platform
        at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
        at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:384)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:228)
        at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:170)
        at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2389)
        at java.base/java.lang.Runtime.load0(Runtime.java:755)
        at java.base/java.lang.System.load(System.java:1953)
        at com.formdev.flatlaf.util.NativeLibrary.loadLibraryFromJar(NativeLibrary.java:108)
        at com.formdev.flatlaf.util.NativeLibrary.<init>(NativeLibrary.java:59)
        at com.formdev.flatlaf.ui.FlatWindowsNativeWindowBorder.getInstance(FlatWindowsNativeWindowBorder.java:113)
        at com.formdev.flatlaf.ui.FlatNativeWindowBorder.initialize(FlatNativeWindowBorder.java:274)
        at com.formdev.flatlaf.ui.FlatNativeWindowBorder.isSupported(FlatNativeWindowBorder.java:68)
        at com.formdev.flatlaf.ui.FlatNativeWindowBorder.install(FlatNativeWindowBorder.java:76)
        at com.formdev.flatlaf.ui.FlatRootPaneUI.installNativeWindowBorder(FlatRootPaneUI.java:205)
        at com.formdev.flatlaf.ui.FlatRootPaneUI.installUI(FlatRootPaneUI.java:105)
        at java.desktop/javax.swing.JComponent.setUI(JComponent.java:730)
        at java.desktop/javax.swing.JRootPane.setUI(JRootPane.java:445)
        at java.desktop/javax.swing.JRootPane.updateUI(JRootPane.java:455)
        at java.desktop/javax.swing.JRootPane.<init>(JRootPane.java:349)
        at java.desktop/javax.swing.JWindow.createRootPane(JWindow.java:274)
        at java.desktop/javax.swing.JWindow.windowInit(JWindow.java:262)
        at java.desktop/javax.swing.JWindow.<init>(JWindow.java:189)

```
So the question is can I ignore this? There seems to be no DLL available for this platform. Is this planned?

**Comment by DevCharly** [comment]
Thanks for reporting.

> So the question is can I ignore this? 

Don't think so. Or does the application work?

Please disable the Flatlaf window decorations (that uses the DLL), via command line option: (see [FlatLaf System Properties](https://www.formdev.com/flatlaf/system-properties/))
~~~
-Dflatlaf.useWindowDecorations=false
~~~

Then it should work without exception.

> There seems to be no DLL available for this platform. Is this planned?

Not at the moment. Simply because I don't have a machine to build and test it. 
Also don't know whether GitHub Actions, which are used to build release DLLs, support Windows ARM.
Maybe it is possible to build a ARM DLL on a x86 machine? Thanks.

Anyway I'm very interested in at least avoiding this exception, but I don't know how to detect whether running on ARM CPU.
Could you please run following program, which dumps Java's system properties, and post the output here?

~~~java
import java.util.Properties;

public class DumpSystemProperties
{
    public static void main( String[] args ) {
        Properties properties = System.getProperties();
        properties.keySet().stream().sorted().forEach( key -> {
            System.out.println( key + " = '" + properties.getProperty( (String) key ) + "'" );
        });
    }
}
~~~

**Comment by xanderberkhout** [comment]
The application works as I don't know what the Window decoration does exactly.

os.arch=aarch64

This means you are running on ARM 64-bit.

**Comment by DevCharly** [comment]
Thanks for the info.

> The application works as I don't know what the Window decoration does exactly.

This enables menu bar embedded into window title bar.
And also the full dark window title bar.

> I tried to run my application on a Mac with M1 processor running Windows 11 in a virtual machine in Parallels.

Many thanks for this hint. Have an M1 Mac. 😄 

A fix for the exception is in latest `2.0-SNAPSHOT` build: https://github.com/JFormDesigner/FlatLaf#snapshots

Would be great if you could try it out.
Hope it works. Thanks.

**Comment by xberkhout** [comment]
I don't see an exception anymore. I have now installed Eclipse on Windows 11 Arm64 so I can debug faster.

**Comment by xanderberkhout** [comment]
It seems the original error is now solved with ARM64 bit Windows, but happens now on Windows x86, 32-bit on the 2.0rc1.

The program continues so it is not a big problem, but it is ugly.

**Comment by DevCharly** [comment]
> ..., but happens now on Windows x86, 32-bit on the 2.0rc1.

What exactly do you mean.
Is there now a `UnsatisfiedLinkError` on 32bit Windows?

What Java version / vendor do you use?

Just tried with a 32bit JRE (17.0.1 from adoption.net) on Windows 10 64bit and it works fine.

**Comment by xanderberkhout** [comment]
yes, there was an UnsatisfiedLinkError with Bellsoft 17.0.1 on Windows 10 32-bit

**Comment by DevCharly** [comment]
Tried Bellsoft 17.0.1 32bit JRE and it works fine (on Windows 10 64bit; don't have 32bit Windows anymore)

What it the value of system property `os.arch`?
And it works with FlatLaf 1.6.x?
Can you post the exception?

**Comment by xanderberkhout** [comment]
It was not my own computer, so I cannot check directly. So I will try to find a virtual machine with Windows 10 32-bit.

**Comment by xanderberkhout** [comment]
I have downloaded Windows 10 32-bit in Parallels and you are right. There is no issue with FlatLaf on Windows 32-bit in this virtual machine. I think there is an issue with the other computer and not with FlatLaf. So sorry to bother you.

**Comment by syoon2** [comment]
I use a Windows on ARM machine (Surface Pro X, to be specific) and tried running a FlatLaf application on it. Didn't encounter any crash or error with the latest stable (v3.1.1), but embedded menu bar seems unavailable.

> Also don't know whether GitHub Actions, which are used to build release DLLs, support Windows ARM.
> Maybe it is possible to build a ARM DLL on a x86 machine? Thanks.

AFAIK GitHub Actions does not offer a Windows on ARM VM. Fortunately Visual C++ provides cross compiling and an ARM DLL can be built on a x86 machine (see #707 for PoC).

I managed to compile and build FlatLaf with an ARM DLL, but even in this version, the embedded menu bar does not work right away (try building & running flatlaf-demo if you need a repro) - not quite sure what the cause of this would be though.

**Comment by DevCharly** [comment]
@syoon2 thanks for working on this issue 👍 

You need to modify these lines:

https://github.com/JFormDesigner/FlatLaf/blob/b47ed94f4068735d65e31fb763a7be6b08cc189e/flatlaf-core/src/main/java/com/formdev/flatlaf/ui/FlatNativeLibrary.java#L58

and

https://github.com/JFormDesigner/FlatLaf/blob/b47ed94f4068735d65e31fb763a7be6b08cc189e/flatlaf-core/src/main/java/com/formdev/flatlaf/ui/FlatNativeLibrary.java#L61

and

https://github.com/JFormDesigner/FlatLaf/blob/b47ed94f4068735d65e31fb763a7be6b08cc189e/flatlaf-core/src/main/java/com/formdev/flatlaf/ui/FlatWindowsNativeWindowBorder.java#L92

BTW please do no commit the ARM DLL to git. I'll do this when merging. I need to digitally sign the DLL.

**Comment by syoon2** [comment]
@DevCharly Yep, #707 is now updated with the edits to the lines you mentioned and ready for review :-)
I tried recompiling and building flatlaf-demo, and the embedded menu bar works flawlessly both on x86_64 and aarch64 :tada:

**Comment by DevCharly** [comment]
Merged in latest `3.2-SNAPSHOT`: https://github.com/JFormDesigner/FlatLaf#snapshots

Could you please test this build and let me know whether it works? Thanks.

**Comment by syoon2** [comment]
Just tested with a Surface Pro X - seems to work as expected :)
